### ISSUE TITLE
Port Verse block to mobile

### ISSUE THREAD

**Issue by guarani** [wordpress-mobile/gutenberg-mobile#1886]
### Description
The aim of this task is to implement the [verse block](https://en.support.wordpress.com/wordpress-editor/blocks/verse-block/) on the mobile Gutenberg editor. The verse block is characterized by its support for line-breaks and ability to render arbitrary whitespace. 

### Considerations
The features to implement on mobile may be a subset of those found in Gutenberg web.

### Subtasks

The subtasks that make up the verse block port are divided into those that are blocking, non-blocking, and undecided (still under investigation).

#### Blocking
- [x] https://github.com/wordpress-mobile/gutenberg-mobile/issues/2037 Enable alignment toolbar for verse block (**pending Aztec Android PR, see issue for details)**
- [x] https://github.com/wordpress-mobile/gutenberg-mobile/issues/1966 Bolding text using the toolbar causes whitespace to be lost
- [x] https://github.com/wordpress-mobile/gutenberg-mobile/pull/2185 Take the verse block out from behind the `devOnly` flag

#### Non-blocking
- [ ] Style the verse block
- [ ] https://github.com/wordpress-mobile/WordPress-iOS/issues/14008 Render line breaks properly on the post list screen on iOS
- [ ] https://github.com/wordpress-mobile/gutenberg-mobile/issues/2193 Use newlines characters instead of `<br>` tags for new lines in pre-based blocks

### Additional Information

Support verse blocks in the Reader https://github.com/wordpress-mobile/WordPress-Android/issues/11299, https://github.com/wordpress-mobile/WordPress-iOS/issues/13452 (these are not Gutenberg per se, just tracking their progress here)

**Comment by guarani** [comment]
### Verse Block vs. Paragraph Block on Gutenberg Web

To help decide which features of the verse block to support on mobile, this is a feature list and comparison between verse and paragraph blocks on Gutenberg web.

Feature Support | Verse  | Paragraph | Comments
| -------------| ------------- | ------------- | ------------- |
| Line-break | ✅| ⚠️ | Hit Return in the verse block, or Shift+Return in the paragraph block  |
| Whitespace | ✅ | ⚠️ | The paragraph block strips whitespace |
| Bold/Italic/Strikethrough | ✅ | ✅ | - |
| Justify text | ✅ | ✅ | - |
| Uppercase text | ✅ | ✅ | - |
| Inline code | ✅ | ✅ | - |
| Additional CSS classes | ✅ | ✅ | - | 
| Hyperlinks | ✅| ✅| - |
| Inline images | ✅| ✅| - |
| Transform to paragraph block | ✅| not applicable | - |
| Transform to group block | ✅| ✅ | - |

**Comment by iamthomasbishop** [comment]
@guarani Thanks for jumping on this block! Stoked to see we're on it 😄 

I took a look at how the web handles design/styling, and it largely depends on the active theme style.

### Using Twenty Twenty theme

Styled similarly to **Preformatted** block.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/1200432/74351990-7d7fce00-4d7d-11ea-8973-ed9e17acc0b6.png">

### Using Maywood theme

Styled like a Paragraph, but with additional padding.

<img width="500" alt="image" src="https://user-images.githubusercontent.com/1200432/74352212-cc2d6800-4d7d-11ea-972e-820f4591423e.png">

### Using Twenty Nineteen

Styled like a paragraph, no additional padding.

<img width="777" alt="image" src="https://user-images.githubusercontent.com/1200432/74352455-23333d00-4d7e-11ea-8561-d582f7c42fb8.png">

## Proposal

In terms of styling, I propose that we should follow the Twenty Nineteen model. In other words:

* No additional padding
* Text styled like a paragraph
* No borders
* No background

**Comment by guarani** [comment]
Thanks for looking at the design, @iamthomasbishop!

On a slightly different topic, I noticed that like other blocks, such as Preformatted and Code, the Verse block is just shown as pre-formatted text when displayed in the mobile Reader. It's out of scope here – but something I think is important to note.


Web Browser | Reader (Android) | Reader (iOS)
-----|-----------------|-----------
<img width="200" src="https://user-images.githubusercontent.com/1898325/74382512-56042200-4dcc-11ea-9a1d-0bbc9c39163a.PNG"> | <img width="200" src="https://user-images.githubusercontent.com/1898325/74382734-b1361480-4dcc-11ea-84ee-5d77c6af6233.png"> | <img width="200" src="https://user-images.githubusercontent.com/1898325/74382519-5997a900-4dcc-11ea-8fa8-9c2c5090384b.PNG">

**Comment by iamthomasbishop** [comment]
That’s a good reminder regarding Reader, @guarani! We should sync at some point with the other team that has Reader things in the apps in their view (perhaps @frosty can advise 🙂).

In terms of Reader styling, I’d suggest the Verse should look the same as I suggested for the editor (similar to paragraph — same font, size, no bg/padding, etc.), whereas **code** and **pre** should look similar to each other.

**Comment by frosty** [comment]
Thanks for the ping! I think this definitely belongs in the iOS / Android repos – @guarani please could you open some issues for this?

**Comment by guarani** [comment]
Adding `verse` to the [list of blocks to register in index.native.js](https://github.com/WordPress/gutenberg/blob/95be29428a6ee23649e46136323c5f0f4c6b540a/packages/block-library/src/index.native.js#L135) was all that was necessary to see it on mobile.

---

### Verse Block Demo

Here are some screenshots of the verse block running on iOS and Android.

| | Block Selector | Verse Block |
|--|--|--|
| iOS | <img width="150" src="https://user-images.githubusercontent.com/1898325/74557953-152d1a00-4f40-11ea-91ec-1d046a8ad19e.png"> | <img width="150" src="https://user-images.githubusercontent.com/1898325/74562766-3a268a80-4f4a-11ea-8764-6bded3d323c7.png"> |
| Android | <img width="150" src="https://user-images.githubusercontent.com/1898325/74562398-7e655b00-4f49-11ea-869d-b99848acaec5.png"> | <img width="150" src="https://user-images.githubusercontent.com/1898325/74563944-a73b1f80-4f4c-11ea-83b0-d7cdd0482780.png"> |

### Current State of Verse Block Support on Mobile

Here are the results of testing each of the features of the verse block on mobile.

Feature | iOS | Android | Comments
| -------------|-------------| -------------| -------------|
| Soft line-break | ✅ | ⚠️ | Hitting return creates a new line. On iOS, new lines in pasted text are preserved. **In Android, new lines in pasted text are not preserved**. Leading new lines and trailing new lines are preserved.
| Hard line-break | ✅ | ✅ | A hard line-break – or new paragraph – is achieved by creating a new verse. |
| Arbitrary whitespace | ✅ | ✅ | Leading, trailing and mid-sentence whitespace is preserved.
| Bold/Italic/Strikethrough/Hyperlinks | 🛑 | ✅ | **iOS only**: Leading whitespace is stripped from lines when bold, italics, links, or strikethrough are applied.<br><br><img width="200" src="https://user-images.githubusercontent.com/1898325/74564494-f2096700-4f4d-11ea-8384-fa397879df4c.gif"> |
| Justify text | 🛑 | 🛑 | This option is missing from the toolbar when a verse block is chosen. |

### Next Steps

Can the following features – while implemented on the web – be left out of scope on mobile?
- Uppercase text
- Inline code
- Additional CSS classes
- Inline images
- Transform to paragraph block
- Transform to group block

I also propose:
- adding the ability to justify text, as it is already implemented for paragraph blocks
- Preserving new lines when pasting text on Android (as noted above)

Regarding hyperlinks, I have not yet been able to test them in the Gutenberg demo app (through the simulators) because the link options (URL, Link text, new tab option) are not appearing, neither for the verse block nor the paragraph block. I'll have to dig further here.

@hypest, @iamthomasbishop – does this all sound ok to you?

**Comment by iamthomasbishop** [comment]
> Can UI for following features – while implemented on the web – be left out of scope on mobile?

I think they can be left out for this iteration, yes.

Thanks for the update!

**Comment by iamthomasbishop** [comment]
Coming back to this issue, I was looking at the example above that shows indented lines, and it made me wonder if we should show `indent` and `outdent` buttons in the block toolbar like we do on the List block? This would essentially act as the sub for the `tab` key. Obviously this beyond scope for the current block, but it could be a useful addition. For visual example:

<a href="https://user-images.githubusercontent.com/1200432/74770656-46645d80-5252-11ea-86ef-a37d5e75a1bb.jpeg"><img src="https://user-images.githubusercontent.com/1200432/74770656-46645d80-5252-11ea-86ef-a37d5e75a1bb.jpeg" width="100"></a>

Note: We could also do this for the Code block as well (esp. if we want to weigh in on the tabs vs. spaces debate 😜).

**Comment by guarani** [comment]
> Coming back to this issue, I was looking at the example above that shows indented lines, and it made me wonder if we should show indent and outdent buttons in the block toolbar like we do on the List block? This would essentially act as the sub for the tab key. Obviously this beyond scope for the current block, but it could be a useful addition. For visual example:

That's interesting, it hadn't occurred to me. If the indent/outdent button worked not only on single lines, but also on multi-line text selections, I think that would be really useful.

I agree this that indenting is a really useful feature for the code block.

**Comment by mchowning** [comment]
I noticed some weird behavior when adding new lines and quickly typing that I think we should follow-up on before removing the dev flag on the block. 

On Android, I see disappearing letters/words. iOS has that as well but also has seems to sometimes move characters and even have some new characters show up unexpectedly. I found it a bit tricky to reproduce this on WPiOS and Android debug builds (I didn't check a release build). On the Android demo app, however, the issue is relatively easy to reproduce.

_ | WPiOS | WPAndroid | Android demo
--- | --- | --- | ---
Description | disappearing characters, toward the end of the gif you'll also see some new characters show up | disappearing characters | disappearing characters/words
Difficulty to Reproduce | Hardest | Hard | Moderate 
gifs | ![verse_weirdness-ios mov](https://user-images.githubusercontent.com/4656348/74964131-dfc17a00-53e0-11ea-9662-2a4fe6243659.gif) | ![verse_weirdness-wpandroid mp4](https://user-images.githubusercontent.com/4656348/74964152-eea82c80-53e0-11ea-8746-6cbdd0849ad7.gif) | ![verse_weirdness-android-demo mp4](https://user-images.githubusercontent.com/4656348/74964531-93c30500-53e1-11ea-86b8-db757813a2bf.gif)

**Comment by guarani** [comment]
From https://github.com/wordpress-mobile/gutenberg-mobile/issues/1886#issuecomment-586462367
> Fixing the stripped whitespace bug (when bold, italics, links, or strikethrough are applied) on iOS

I've broken this out into a separate bug issue with more context as it affects all blocks, not just the verse block. See https://github.com/wordpress-mobile/gutenberg-mobile/issues/1966

**Comment by guarani** [comment]
Regarding styling (https://github.com/wordpress-mobile/gutenberg-mobile/issues/1886#issuecomment-585276207):

> In terms of styling, I propose that we should follow the Twenty Nineteen model. In other words:
> 
>     No additional padding
>     Text styled like a paragraph
>     No borders
>     No background

I want to use the exact same styles as the paragraph block, but from what I'm seeing, the paragraph block's styling on mobile is by virtue of it using the `<p>` tag and how Aztec interprets that – its styling doesn't come from scss as far as I can tell.

As a test, I changed the verse block to use a `<p>` tag instead of a `<pre>` tag and that achieved the styling I was looking for This doesn't look like a solution though, because this would make the mobile editor output HTML for verse blocks which are fundamentally different from the web.

I could just try to play with the scss properties to make it visually match the paragraph block, or take a deep dive into Aztec to see what styling it's putting on `<p>` tags, but I'm not sure if I'm on the right track. Any thoughts here @hypest and @mchowning ?

**Comment by hypest** [comment]
A quick note:

On the mobile ports of the blocks, we never actually change the _output_ markup of the block. In other words, we never need to change the `save()` function of the block. So, if there's a recommendation here to differentiate the frontend rendering of Verse block, I'd pause to rethink its need. I see that the [current `save()` function of](https://github.com/WordPress/gutenberg/blob/00c92c746d0e2a67edc9ca71ed76fd6be37fa9d5/packages/block-library/src/verse/save.js#L19-L23) the Verse is using a `pre`

**Comment by guarani** [comment]
A recent change on Gutenberg (https://github.com/WordPress/gutenberg/pull/20752) has fixed verse block styling – it's now the same as the paragraph block. Just noticed this today.

You can see in the below screenshots how the verse and paragraph blocks are now styled identically. I'm unclear on how this change works because under the hood the verse is still using the `pre` block (I switched to HTML mode to check this).

| Android | iOS |
| -- | -- |
| <img src="https://user-images.githubusercontent.com/1898325/77793897-4d487200-7041-11ea-9ea1-f9a88a71e883.png"> | <img src="https://user-images.githubusercontent.com/1898325/77794152-c34cd900-7041-11ea-96e1-d4d611763e71.png"> |

**Comment by hypest** [comment]
Aha, that's good news 😄 .

Ofc, having the PRE block look exactly like the paragraph is not desirable since the PRE should have a monospace font. I've opened a ticket for it https://github.com/wordpress-mobile/gutenberg-mobile/issues/2081.

**Comment by guarani** [comment]
@iamthomasbishop, after discussion with @hypest and @SergioEstevao, the "fix" i mentioned above in https://github.com/wordpress-mobile/gutenberg-mobile/issues/1886#issuecomment-605286061 is no longer the case. The verse block uses a `pre` tag under the hood to preserve spaces and line breaks, but Aztec always styles `pre` tags using a monospace font. Also, we can't change the Verse block to use a `p` tag as this would mean spaces and line-breaks are no longer preserved.

The proposal we considered is to leave the style of the verse block alone for now and address it at a later stage. What do you think @iamthomasbishop ?

**Comment by guarani** [comment]
Update to the above comment, we're also proposing leaving the "Bolding text using the toolbar causes whitespace to be lost" subtask out for now, so we can unblock the release of the Verse block which is still only available on development/debug builds.
It's already got its own issue: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1966

**Comment by iamthomasbishop** [comment]
@guarani Apologies on the delay. I hope I'm understanding the issue properly. Is there anything we can latch onto that allows us to style the verse variation of `pre` in order to give it a monospace font?

> we're also proposing leaving the "Bolding text using the toolbar causes whitespace to be lost" subtask out for now, so we can unblock the release of the Verse block

That sounds like a bug that we should fix prior to shipping the block, no?

**Comment by hypest** [comment]
Echoing Thomas above, the whitespace removal seems like a blocker for un-DEVing the block. What's the context behind the thought to not block on that issue?

**Comment by guarani** [comment]
> Is there anything we can latch onto that allows us to style the verse variation of pre in order to give it a monospace font?

It's my understanding that we want to give it a _non-monospace_ font, as per your comment here: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1886#issuecomment-585276207. Please let me know if there's been a misunderstanding. 
If our goal is to give it a monospace font, I think that's already done – you can see what it currently looks like here, https://github.com/wordpress-mobile/gutenberg-mobile/issues/1886#issuecomment-586462367.

> That sounds like a bug that we should fix prior to shipping the block, no?

> Echoing Thomas above, the whitespace removal seems like a blocker for un-DEVing the block. What's the context behind the thought to not block on that issue?

My reasoning is that since this is an existing bug in production (I'm not sure if I made that clear earlier), it's not specific to the verse block and equally affects any other rich text (e.g. pre-formatted and paragraph blocks). This doesn't mean it's not a good time to fix this bug now. Perhaps users of the verse block use styled and indented (leading whitespace) content more often than users of the pre-formatted or paragraph block.

**Comment by iamthomasbishop** [comment]
> It's my understanding that we want to give it a non-monospace font, as per your comment here

Ack yes, you’re right, I had it reversed 🤦‍♂ 

> My reasoning is that since this is an existing bug in production (I'm not sure if I made that clear earlier), it's not specific to the verse block and equally affects any other rich text

Ok, that makes sense. I’d like to fix asap and pause shipping unless this fix will take a long time — in which case we can tackle separately.

> Perhaps users of the verse block use styled and indented (leading whitespace) content more often than users of the pre-formatted or paragraph block.

I’m not sure — purely an assumption, but I’d assume indentation would probably also be common on the `pre` block, but not as common on the `paragraph` block.

**Comment by guarani** [comment]
No worries – I'm doing a bit more debugging on this and will write up my findings on the subissue in question, https://github.com/wordpress-mobile/gutenberg-mobile/issues/1966. I'll ping you here once that's done 👍.

**Comment by guarani** [comment]
@iamthomasbishop, sorry to go back to what we were discussing earlier. 
Could you please confirm if it is acceptable for the verse block to be released using a monospace font in the editor? This won't affect how the block is rendered on the web, it only affects how users see the verse block in the mobile editor. 
I understand we would prefer that the verse block use the same font as the paragraph block in the editor, but in offline discussion with @hypest we concluded that would not be an easy fix.

**Comment by iamthomasbishop** [comment]
> Could you please confirm if it is acceptable for the verse block to be released using a monospace font in the editor?

I think it's acceptable, but hoping we can iterate on it quickly and apply the expected serif font.

**Comment by kspilarski** [comment]
> Coming back to this issue, I was looking at the example above that shows indented lines, and it made me wonder if we should show `indent` and `outdent` buttons in the block toolbar like we do on the List block? This would essentially act as the sub for the `tab` key. Obviously this beyond scope for the current block, but it could be a useful addition. For visual example:
> 
> <img alt="" width="100" src="https://user-images.githubusercontent.com/1200432/74770656-46645d80-5252-11ea-86ef-a37d5e75a1bb.jpeg">
> 
> Note: We could also do this for the Code block as well (esp. if we want to weigh in on the tabs vs. spaces debate 😜).

Related GitHub report:  [Code & SyntaxHighlighting Blocks: Indent / Outdent Option](https://github.com/Automattic/wp-calypso/issues/44131)

**Comment by mchowning** [comment]
👋  @guarani ! Should we close this issue (and maybe creating new issues for things [like indentation](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1886#issuecomment-658138281)) or do you think we would be better off leaving this issue open?

**Comment by guarani** [comment]
This was mostly for tracking the effort needed to ship the first block, so I agree we could close this and add issues for the pending items. I'll do this soon, thanks for flagging this 👍

**Comment by guarani** [comment]
I created a separate issue to track the work that needs to be done to style the verse block like a paragraph block: https://github.com/WordPress/gutenberg/issues/30836

The other non-blockers here all have their own issues already so I think this is safe to close.
